### PR TITLE
inherit fields for an alias from parent adapter

### PR DESCRIPTION
**Pull request by gargcreation1992**
This PR makes the following changes - 
- It introduces a new struct `bidderInfoNullableFields` to maintain any nullable fields from BidderInfo struct. For example - the default value of `Disabled` in BidderInfo struct is always false and not `nil`.
- It may be the case where `Disabled` value of parent adapter is `true` but no value is set for an aliased adapter(default is false). In this case, we cannot just outright replace `Disabled` value of alias adapter with parent. Consider the case where aliased adapter `Disabled` value is actually set.
- To handle cases like above, we have introduced `bidderInfoNullableFields` struct.
- The approach is to maintain a map of bidderName to `bidderInfoNullableFields`. 
- Loop information from that map to determine if `Disabled` is actually set in aliased adapter or not since pointer values are defaulted to nil in golang. We do not have more than 20 aliases as of today so this loop is not expensive.
- This PR just inherits the fields from parent adapter to aliased adapter. In the subsequent PR, we will be handling the cases where nullable fields of BidderInfo struct is set.

### CODE REVIEW

File: config/bidderinfo.go [hunk 1]
```diff
@@ -229,8 +238,8 @@ func processBidderInfos(reader InfoReader, normalizeBidderName func(string) (ope
 		return nil, fmt.Errorf("error loading bidders data")
 	}
 
-	infos := BidderInfos{}
-
+	bidderInfos := BidderInfos{}
+	aliasInfos := aliasInfos{} //maintains map of bidderName to nullable fields in BidderInfo struct
```
Comment (onkarvhanumante) [inline config/bidderinfo.go:242]: how about changing comment to following or something similar

```
// aliasInfos maintains a map of bidderName to nullable fields
// will be used to determine values that are not defined in the bidderInfo yaml
aliasInfos := aliasInfos{} 
```

File: config/bidderinfo.go [hunk 2]
```diff
@@ -243,10 +252,49 @@ func processBidderInfos(reader InfoReader, normalizeBidderName func(string) (ope
...
-	return infos, nil
+	return processBidderAliases(aliasInfos, bidderInfos)
+}
+
+func processBidderAliases(aliasInfos aliasInfos, bidderInfos BidderInfos) (BidderInfos, error) {
+	for bidderName := range aliasInfos {
+		if err := validateAliases(bidderInfos[bidderName], bidderInfos, bidderName); err != nil {
+			return nil, err
+		}
+		aliasBidderInfo := bidderInfos[bidderName]
```
Comment (onkarvhanumante) [inline config/bidderinfo.go:277]: Here, we lookup twice on `bidderInfos` for same bidder name on line 274 and 275. Also look  up on `bidderInfos` map is done using  key from `aliasInfos` map. So should check `aliasBidderInfo` is present before making call to `validateAliases()`




```
aliasBidderInfo, ok := bidderInfos[bidderName]
if !ok {
    // return error 
}

if err := validateAliases(aliasBidderInfo, bidderInfos, bidderName); err != nil {
    return nil, err
}
...
```

File: config/bidderinfo.go [hunk 3]
```diff
@@ -243,10 +252,49 @@ func processBidderInfos(reader InfoReader, normalizeBidderName func(string) (ope
...
+		parentBidderInfo := bidderInfos[aliasBidderInfo.AliasOf]
+		aliasBidderInfo.AppSecret = parentBidderInfo.AppSecret
```
Comment (hhhjort) [inline config/bidderinfo.go:279]: So we will always overwrite all these fields of an alias with the value from the parent? I believe we want an alias to specify a different config than the parent for at least some of these.

File: config/bidderinfo.go [hunk 4]
```diff
@@ -243,10 +252,49 @@ func processBidderInfos(reader InfoReader, normalizeBidderName func(string) (ope
...
+
+			//need to maintain nullable fields from BidderInfo struct into bidderInfoNullableFields
+			//to handle the default values in aliases yaml
+			if len(info.AliasOf) > 0 {
+				aliasInfo := bidderInfoNullableFields{}
+				if err := yaml.Unmarshal(data, &aliasInfo); err != nil {
+					return nil, fmt.Errorf("error parsing config for aliased bidder %s: %v", fileName, err)
+				}
+
+				aliasInfos[string(normalizedBidderName)] = aliasInfo
```
Comment (hhhjort) [inline config/bidderinfo.go:265]: What do we do with the `aliasInfo` struct we are saving here? You pass `aliasInfos` in to `processBidderAliases()`, but that only uses `aliasInfos` to loop over the keys of the map. I don't see where you are accessing the struct you have stored in the map.

File: config/bidderinfo.go [hunk 5]
```diff
@@ -52,6 +52,15 @@ type BidderInfo struct {
 	OpenRTB             *OpenRTBInfo `yaml:"openrtb" mapstructure:"openrtb"`
 }
 
+type bidderInfoNullableFields struct {
```
Comment (SyntaxNode) [inline config/bidderinfo.go:55]: Nitpick: In Go, this should be `NillableFields`

File: config/bidderinfo.go (hunk 1 above)
Comment (SyntaxNode) [reply config/bidderinfo.go:242]: Why the use of the `aliasInfos` type alias? There's no need to copy `BidderInfos` in this regard. Might be easier to include the type directly. The name of `aliasNillableFieldsByBidder` might remove the need for a comment entirely.

File: config/bidderinfo.go (hunk 3 above)
Comment (SyntaxNode) [reply config/bidderinfo.go:279]: There is a PR dependent on this one for conditional overwrites: #3001

File: config/bidderinfo.go (hunk 4 above)
Comment (gargcreation1992) [reply config/bidderinfo.go:265]: There is a PR dependent on this one for conditional overwrites: https://github.com/prebid/prebid-server/pull/3001

File: config/bidderinfo.go (hunk 2 above)
Comment (gargcreation1992) [reply config/bidderinfo.go]: https://github.com/prebid/prebid-server/pull/3000/commits/92211a53da0a3a1dcf80c64c792f36265d88ffcb

File: config/bidderinfo.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gargcreation1992) [reply config/bidderinfo.go]: changed the name of the structs to be more descriptive so removed the comments - https://github.com/prebid/prebid-server/pull/3000/commits/c6be49c2ffcda9511cbe40163240a27aeec1bf52

Comment (gargcreation1992) [reply config/bidderinfo.go]: https://github.com/prebid/prebid-server/pull/3000/commits/c6be49c2ffcda9511cbe40163240a27aeec1bf52

**Review by onkarvhanumante — APPROVED** [review]

File: config/bidderinfo.go (hunk 3 above)
Comment (hhhjort) [reply config/bidderinfo.go:280]: That would explain it.

**Pull request merged**